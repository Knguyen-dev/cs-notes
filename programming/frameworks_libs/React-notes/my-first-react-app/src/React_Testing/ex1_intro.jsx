import { useState } from "react"
const App = () => {
	const [heading, setHeading] = useState("Magnificent Monkeys!")

	const clickHandler = () => {
		setHeading("Radical Rhinos")
	}

	return (
		<>
			<button type="button" onClick={clickHandler}>
				Change Text
			</button>
			<h1>{heading}</h1>
		</>
	)
}

export { App }
