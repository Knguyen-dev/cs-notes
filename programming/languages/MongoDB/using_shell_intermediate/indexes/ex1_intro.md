# Indexes in MongoDB
- Ordered data-structures that allow for fast data retrieval.
  Querying is a lot more efficiently when MongoDB doesn't have to 
  examine the entire collection of documents. Indexes support 
  equality matches and ranged based queries, and return sorted results.

  1. Without Indexes: MongoDB has to read all documents 
    in collection before returnnig our result.
  2. With Indexes: MongoDB only fetches documents recorded
    by the index. As a result, MongoDB becomes somewhat faster.

# Costs of Indexing:
- The drawback is write performance. If you update the collection, be sure 
  to update the indexes as well. And when you have a lot of stuff in the collection, it's going to be 
  quite a bit of work updating the indexes. So when using indexes, we improve
  read/query performance at the cost of write performance.

# Types of indexes:
1. Single field: Indexes that track one field only
2. Compound: Indexes that track multiple fields
3. Multikey: Name for when there are indexes that operate on an 
  array field.





- Situation: When doing db.collection("books").find({"rating":10}), 
  MongoDB has to check every document. Now this can get a little inefficient 
  if you have thousands of documents. 

- Solution: Make an index for whatever field you need to query. This 
  lists out everything you need to look up. You'd have a list 
  called "book_rating" index, which would be a list of book ratings. However
  each entry in that list, would also have a pointer to the book (in the 'books' collection)
  to the book the rating is associated with. 

  So now when the query is based on rating, MongoDB scansthe book_rating
  index instead of looking through all of those documents, 
  to find all 10s in a much faster way. Then it would link back and return
  the documents associated with those 10s.

  book_rating index = [1,5,10,10,9]; We have two 10 ratings here, and essentially 
  these 10 ratings would have pointers back to the book documents they're associated 
  with.

NOTE: Indexes are based on one field value only.

# When to use indexes:
- Drawbacks: You don't always need to use indexes. I mean you're probably not going to 
  create an index for something like 'number of pages'. Also it's somewhat 
  difficult to maintain as if you change your collection, the indexes must match 
  as well. And when you have a lot of stuff in the collection, it's going to be 
  quite a bit of work updating the indexes.
- Use cases: Where you need a specific query that only return a subset of documents.
  Or when you have a very large amount of documents in a collection, where you need
  to sort when you query them.

# Viewing Execution Statistics
1. db.newBooks.find({rating: 8}).explain("executionStats");
- We see some output, and we go to the 'executionStats' field. 
  Here it tells us the total number of documents it looked at and 
  the number of documents returned. In my case it returned 3 docs, 
  and looked at 8 (the total number of docs in my collection). This
  makes sense, as MongoDB is examining every last document to ensure 
  it isn't missing any that meet our 'rating'.


# How to make indices (in the shell)
- Lets create an index based on the rating value. 
1. db.newBooks.createIndex({rating: 8}); We created an index
  for the rating of 8. MongoDB scanned our newBooks collection
  and indexed/recorded all of the documents with the rating of 8.
  Now when we want to find documents in newBooks, with a rating of 8,
  it's going to do it much quicker via that index. For example,
  let's say there were only two books with a rating=8, now when
  it uses the index we made, the query only examines 2 documents
  instead of querying the entire database like before. That's because
  it only looked at the documents listed in our index.

2. db.newBooks.getIndexes(); Get the indexes created for the newBooks
  collection. You'll see there's an "_id_" one that was auto-created by
  MongoDB, but now there's that "rating_8" index we created.

3. db.books.dropIndex("rating_8"); deletes an index, so here
  we delete that "rating_8" index that we created.

