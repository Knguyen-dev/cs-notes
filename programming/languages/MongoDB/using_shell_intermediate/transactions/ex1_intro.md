# Transactions
- Situation: Let's say at dinner friendA offers to pay the bill, and
  the friend B will just transfer money friend A's account as a way
  to pay friend A for paying the bill at this time. In this case, it'd
  be bad if one of the operations didn't work such as, friend A pays the 
  bill, but friend B doesn't compensate A. We need a way to make sure either
  both of these operations happen, or the whole thing is cancelled.

- Idea of transactions: Any time database operations transfer a value from one record to 
another, we need to guarantee these operation happen all together or not at all
to guarantee the integrity of our data. ACID transactions are a group of database
operations that will be completed as a unit or not at all.


# ACID
1. Atomicity: All operations will either succeed or fail together
2. Consistency: All changes are consistent with database constraints.
  For example, if an account balance can't go below zero, then a transaction
  will fail if one of its operations violates this condition.
3. Isolation: Multiple transactions can happen simulanteously without
  affecting the outcomes of other transactions. 
4. Durability: All changes made by operations will persist, even in
  the event of power or hardware failure. Regardless, after a transaction
  the data should persist, and data is never lost.


# When to use ACID transactions:
- Should be used in scenarios that involve the transfer 
  of a value from one record to another. Some examples include:
1. A mobile payment app where we've moving funds.
2. Systems that track inventory such as online shopping cart.