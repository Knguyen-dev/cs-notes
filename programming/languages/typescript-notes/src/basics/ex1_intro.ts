/*
+ Basic intro:
- Defining a data-type for a parameter. Here
we can define the type of `myStr` as a string.
*/
function reverse(myStr: string) {
	return myStr.split("").reverse().join("");
}

/*
- You can even define custom types. So if we had a MenuItem 
  object and you're passing it into a function, you could add
  a custom type to ensure that other developers know what 
  'menuItemObj' is and what it looks like.
*/
interface MenuItem {
	title: string;
	cost: number;
}

function addToCart(item: MenuItem) {
	return `Added ${item.title} for $${item.cost}`;
}
