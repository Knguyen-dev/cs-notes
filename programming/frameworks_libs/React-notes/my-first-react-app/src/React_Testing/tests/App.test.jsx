/*
+ Importances of Querying By Role: The importance is that by querying ByRole ensures that our UI is accessible to 
    everyone whether they use a mouse or other assistive technologies.

+ Types of queries
1. getBy: Returns matching node. Returns error if nothing is found or if there are multiple matches
2. queryBy: Returns matching node. Returns null if no match. Returns an error if there are multiple matches.
3. findBy: Returns a Promise which resolves if element is found, and rejected if 
    nothing is found or multiple elements are found after a 1 second timeout.

- NOTE: If you're wanting to select multiple elements use getAllBy, queryAllBy, or findAllBy.
1. getAllBy: Returns an array of nodes, or throws an error if nothing matches
2. queryAllBy: Returns array of nodes, or empty array if nothing matches
3. findAllBy: Returns promise that resolves into array of elements, or rejects 
    when nothing is found after 1 second timeout.


Priority (Best options to query the tree with):
1. getByRole: Queries all elements in accessibility tree, which is a tree based on the DOM tree but for accessibility 
    related information. So to query for DOM nodes, we query by their accessibility 'roles' instead.
    Use the 'name' option to filter elements by their accessible name. Typically should be the goto pick, and
    here's a list of roles to help you with querying.

    NOTE: 
    1. Accessible name is text dictated by aria-label, aria-labelled-by, or the text content 
        of the element. For buttons, since aria-label and aria-labelledby are usually absent, 
        the browser and accessibility standards use the text content as the name. 

    List of roles: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#Roles
    CheatSheet for queries: https://testing-library.com/docs/dom-testing-library/cheatsheet/

2. getByLabelText: Really good for form fields since those have labels 
3. getByPlaceholderText: Not supposed to be a substitute for getByLabelText, but if you don't 
    have a label in the DOM, then this is a good alternative.
4. getByText: Can be used for finding non-interactive elements usch as divs, spans, paragraphs, etc.
5. getByDisplayValue: Used for querying by the current value of a form element
6. getByAltText: Queries for alt text.

NOTE: We use "screen" because it's easier, and without it we'd need to use a container.
    Base queries from DOM Testing Library require you give a container, and most frameworks 
    do this for you. Main takeaway is using screen is recommended.
    For example:

- With Screen:
const inputNode = screen.getByLabelText("Username")

- Without Screen:
const container = document.querySelector("#app") // query for our app node
const labelNode = getByLabelText(container, "Username") // query for a label with text "Username"


+ Simulating User Events: Of course live user feedback/interaction is good, but building
    confidence through tests is good too


*/

import { describe, it, expect } from "vitest"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { App } from "../ex1_intro"

/*
Example 1: In this test, we're seeing testing our App class actually rendered a heading markup with 
1. Render our app on our testing DOM
2. One way is to query the heading node and get its text content, and see if its text matches. In order to 
    test if our heading rendered correctly.
3. The other way, using name, is to see if there exists a heading node with name or text 'Our First Tst"est

describe("App component", () => {
	it("Renders correct heading", () => {
		render(<App />)
		// expect(screen.getByRole("heading").textContent).toMatch(/our first test/i)
		expect(
			screen.getByRole("heading", {
				name: /Our First Test/i,
			})
		)
	})
})
*/

/*
- Example 2: Testing user events
1. 

*/

describe("App Component", () => {
	/*

    + Snapshot testing: Comparing our rendered component with an associated snapshot file. When
        running this test, an associated snapshot file rendered, which captured the markup of the App component
        on our virtual DOM. A snapshot test is testing if our markup output is matching the previous snapshot of
        our markup.

    - render(<App/>):
        1. Renders app component in a virtual DOM for testing purposes.
        2. Render() also returns an object with a property 'container' which is a reference to
            the root DOM node that contains the App component's markup. So container is our current DOM output.
    - toMatchSnapshot(): Takes a snapshot of the currently rendered output in our virtual DOM, and compares it something. 
        If there is no previous snapshot, it takes a snapshot of our current DOM.
        
        
    - How it works: So on our initial run, no previous snapshot so it takes a snapshot of the current DOM. Then 
        we compared container to match that said snapshot of our markup. This should test if the markup of 
        the snapshot in the __snapshots__ folder, matches 'container', which passes. Then on subsequent runs,
        we continue to compare our container against previously stored snapshot 
 

    - Importance: May seem dumb, but the idea is checking "hey is this component rendering consistently with the 
        same information?". Snapshot testing is more pure ocmponents that render UI based on props. Things 
        that don't rely on complex logic or fetching varying data. Here it passes because we intended and designed
        our component to be pure, and not render varying markup everytime.


    NOTE: If you markup for your component while snapshot testing, in this case it'll fail because it's testing if 
        our current newly changed markup is the same as the old snapshot taken in the past. However, you'll get 
        an option to update the snapshot to match the new markup so that tests run well.
    */
	it("renders magnificent monkeys", () => {
		const { container } = render(<App />)
		expect(container).toMatchSnapshot()
	})

	/*
    + Simulating User Events: Here we're simulating a click event. Then we check if the heading's text
        has changed with 'toMatch', which is one of the various assertions we could have used.

    NOTE: 
        1. Callback function is asynchronous because user.click() simulates the asynchronous nature
        of user interaction.

        2. always do userEvent.setup() before rendering the component that you're testing. 

        3. React Testing Library unmounts the rendered components after every test. 
            This is why we're rendering the App on every test case.
    */
	it("render radical rhinos after button click", async () => {
		const user = userEvent.setup()
		render(<App />)
		const button = screen.getByRole("button", { name: "Change Text" })
		await user.click(button)
		expect(screen.getByRole("heading").textContent).toMatch(/radical rhinos/i)
	})
})

/*
- Extra for simulating events: You can also simulate keyboard events if needed.

1. simulates user pressing certain keys 
2. Using brackets. Like the first, we aren't keeping any keys pressed, so we are lifting Shift
    before pressing f.
- More can be found on Docs: https://testing-library.com/docs/user-event/intro


const user = userEvent.setup()
await user.keyboard('foo'); user presses f, o, o
await user.keyboard('{Shift}{f}{o}{o}'); Shift, f, o, o




*/
