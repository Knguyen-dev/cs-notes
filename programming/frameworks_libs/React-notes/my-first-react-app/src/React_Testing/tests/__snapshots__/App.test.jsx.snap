// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > renders magnificent monkeys 1`] = `
<div>
  <button
    type="button"
  >
    Change Text
  </button>
  <h1>
    Magnificent Monkeys!
  </h1>
</div>
`;
